general/occa.hpp

```diff
@@ -15,29 +15,68 @@
 #include "../config/config.hpp"
 
 #ifdef MFEM_USE_OCCA
+#include "mem_manager.hpp"
+#include "device.hpp"
 #include <occa.hpp>
 
-#if defined(MFEM_USE_CUDA) && OCCA_CUDA_ENABLED
-#include <occa/modes/cuda/utils.hpp>
-#endif
+namespace mfem
+{
 
-typedef occa::device OccaDevice;
-typedef occa::memory OccaMemory;
+/// Return the default occa::device used by MFEM.
+occa::device &OccaDev();
 
-#else // MFEM_USE_OCCA
+/// Wrap a pointer as occa::memory with the default occa::device used by MFEM.
+/** It is assumed that @a ptr is suitable for use with the current mfem::Device
+    configuration. */
+occa::memory OccaMemoryWrap(void *ptr, std::size_t bytes);
 
-typedef void* OccaDevice;
-typedef void* OccaMemory;
+/** @brief Wrap a Memory object as occa::memory for read only access with the
+    mfem::Device MemoryClass. The returned occa::memory is associated with the
+    default occa::device used by MFEM. */
+template <typename T>
+const occa::memory OccaMemoryRead(const Memory<T> &mem, size_t size)
+{
+   mem.SetExecFlag(true);
+   const void *ptr = mem.ReadAccess(Device::GetMemoryClass(), size);
+   return OccaMemoryWrap(const_cast<void *>(ptr), size*sizeof(T));
+}
 
-#endif // MFEM_USE_OCCA
+/** @brief Wrap a Memory object as occa::memory for write only access with the
+    mfem::Device MemoryClass. The returned occa::memory is associated with the
+    default occa::device used by MFEM. */
+template <typename T>
+occa::memory OccaMemoryWrite(Memory<T> &mem, size_t size)
+{
+   mem.SetExecFlag(true);
+   return OccaMemoryWrap(mem.WriteAccess(Device::GetMemoryClass(), size),
+                         size*sizeof(T));
+}
 
-namespace mfem
+/** @brief Wrap a Memory object as occa::memory for read-write access with the
+    mfem::Device MemoryClass. The returned occa::memory is associated with the
+    default occa::device used by MFEM. */
+template <typename T>
+occa::memory OccaMemoryReadWrite(Memory<T> &mem, size_t size)
 {
+   mem.SetExecFlag(true);
+   return OccaMemoryWrap(mem.ReadWriteAccess(Device::GetMemoryClass(), size),
+                         size*sizeof(T));
+}
+
 
-// Function called when the pointer 'a' needs to be passed to an OCCA kernel.
-OccaMemory OccaPtr(const void *a);
-OccaDevice OccaDev();
+/** @brief Function that determines if an OCCA kernel should be used, based on
+    the current mfem::Device configuration. */
+inline bool DeviceUseOcca()
```

Comment:
The function name looks like an action
```suggestion
inline bool DeviceCanUseOcca()
```

Reply:
Good point. I'll change it.